File: priv/hornbeam_asgi_loop.py
# Copyright 2026 Benoit Chesneau
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""ASGI handler using Cowboy loop handler with direct message passing.

Simple design:
- Request body: read from Erlang channel in receive()
- Response: send via erlang.send() directly to Cowboy handler
- No buffering, no extra tasks
"""

from typing import Callable, Optional

try:
    import erlang
    from erlang import ByteChannel, ByteChannelClosed
    HAS_ERLANG = True
    _erlang_send = erlang.send
except ImportError:
    HAS_ERLANG = False
    erlang = None
    _erlang_send = None
    ByteChannel = None
    ByteChannelClosed = Exception


class ASGIProtocol:
    """ASGI handler using Erlang channels for request body.

    Simple design: read from channel directly in receive(), no buffering.
    """

    def __init__(self, caller_pid, app: Callable, scope: dict,
                 req_channel: Optional[ByteChannel]):
        self._caller_pid = caller_pid
        self._app = app
        self._scope = scope
        self._req_channel = req_channel  # None for no-body requests
        self._send_fn = _erlang_send

        # Response state
        self._response_started = False
        self._response_finished = False

    # =========================================================================
    # ASGI interface
    # =========================================================================

    async def receive(self) -> dict:
        """ASGI receive callable - reads directly from channel."""
        # No body case
        if self._req_channel is None:
            return {'type': 'http.request', 'body': b'', 'more_body': False}

        # Read from channel
        try:
            chunk = await self._req_channel.async_receive_bytes()
            return {
                'type': 'http.request',
                'body': chunk if chunk else b'',
                'more_body': True,
            }
        except ByteChannelClosed:
            return {'type': 'http.request', 'body': b'', 'more_body': False}

    async def send(self, message: dict) -> None:
        """ASGI send callable.

        Uses erlang.send() directly for all response data - no channel overhead.
        """
        if self._response_finished:
            raise RuntimeError("Response already completed")

        msg_type = message.get('type', '')

        if msg_type == 'http.response.start':
            if self._response_started:
                raise RuntimeError("http.response.start already sent")

            status = message.get('status', 200)
            headers = message.get('headers', [])
            self._send_fn(self._caller_pid, (b'headers', status, headers))
            self._response_started = True

        elif msg_type == 'http.response.body':
            if not self._response_started:
                raise RuntimeError("http.response.start must be sent first")

            body = message.get('body', b'')
            if isinstance(body, str):
                body = body.encode('utf-8')

            more_body = message.get('more_body', False)

            # Send body directly via erlang.send - no channel overhead
            self._send_fn(self._caller_pid, (b'body', body, more_body))

            if not more_body:
                self._response_finished = True

        elif msg_type == 'http.response.informational':
            status = message.get('status', 100)
            headers = message.get('headers', [])
            if status == 103:
                self._send_fn(self._caller_pid, (b'early_hints', headers))

        elif msg_type == 'http.disconnect':
            self._response_finished = True

    async def run(self):
        """Run the ASGI application."""
        try:
            await self._app(self._scope, self.receive, self.send)

            # Ensure response is completed
            if not self._response_finished:
                if not self._response_started:
                    self._send_fn(self._caller_pid, (b'headers', 500, []))
                self._send_fn(self._caller_pid, (b'body', b'', False))

        except Exception as e:
            self._send_fn(self._caller_pid, (b'error', str(e).encode('utf-8')))


# =============================================================================
# Entry point
# =============================================================================

async def handle_asgi_loop(caller_pid, app_module: str, app_callable: str,
                           scope: dict, req_body_ch):
    """Handle ASGI request using Protocol pattern.

    Entry point called from hornbeam_asgi_loop.erl.
    Response is sent directly via erlang.send() - no channel needed.
    """
    if not HAS_ERLANG:
        return

    # Wrap request channel (may be 'empty' atom for no-body requests)
    if req_body_ch == 'empty' or req_body_ch == b'empty':
        req_channel = None
    else:
        req_channel = ByteChannel(req_body_ch)

    # Get app
    app = _get_app(app_module, app_callable)

    # Wrap scope['state'] if present
    if 'state' in scope:
        mount_id = scope.get('mount_id')
        scope['state'] = _MutableStateProxy(scope['state'], mount_id)

    # Create and run protocol
    protocol = ASGIProtocol(caller_pid, app, scope, req_channel)
    await protocol.run()


# =============================================================================
# Helpers
# =============================================================================

_preloaded_app: Callable = None
_preloaded_key: tuple = None


def preload_app(app_module: str, app_callable: str) -> bytes:
    """Preload ASGI application at startup."""
    global _preloaded_app, _preloaded_key

    import importlib
    module = importlib.import_module(app_module)
    app = getattr(module, app_callable)

    _preloaded_app = app
    _preloaded_key = (app_module, app_callable)

    return b'ok'


def _get_app(module_name: str, callable_name: str) -> Callable:
    """Get ASGI application."""
    global _preloaded_app, _preloaded_key

    if _preloaded_key == (module_name, callable_name):
        return _preloaded_app

    import importlib
    module = importlib.import_module(module_name)
    return getattr(module, callable_name)


class _MutableStateProxy(dict):
    """Dict that syncs mutations to Erlang ETS."""

    __slots__ = ('_mount_id', '_lifespan_pid')

    def __init__(self, initial_state: dict, mount_id=None):
        super().__init__(initial_state)
        self._mount_id = mount_id
        self._lifespan_pid = None
        if HAS_ERLANG:
            try:
                self._lifespan_pid = erlang.whereis("hornbeam_lifespan")
            except Exception:
                pass

    def __setitem__(self, key, value):
        super().__setitem__(key, value)
        if self._lifespan_pid is not None:
            try:
                if self._mount_id is not None:
                    _erlang_send(self._lifespan_pid,
                                (b'update_state', self._mount_id, key, value))
                else:
                    _erlang_send(self._lifespan_pid, (b'update_state', key, value))
            except Exception:
                pass

File: src/hornbeam_asgi_loop.erl
%% Copyright 2026 Benoit Chesneau
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

%%% @doc ASGI handler using Cowboy loop handler with push/pull body streaming.
%%%
%%% This module implements ASGI request handling using Cowboy's async body
%%% reading via loop handlers. Instead of spawning a pump process, it uses
%%% cowboy_req:cast to receive body chunks as messages, which are then
%%% pushed to Python via channel.
%%%
%%% Architecture:
%%% - Cowboy sends {request_body, Ref, nofin|fin, Data} messages
%%% - Loop handler info/3 pushes chunks to Python channel
%%% - Python buffers chunks and signals via asyncio.Event
%%% - ASGI receive() pulls from buffer
%%%
%%% Benefits over pump process approach:
%%% - No extra process spawn per request
%%% - Event-driven data flow
%%% - Natural backpressure via channel
%%% - Better integration with Cowboy's internals
%%%
%%% @end
-module(hornbeam_asgi_loop).

-behaviour(cowboy_loop).

-export([init/2, info/3, terminate/3]).

%% Internal state
-record(state, {
    req_info,
    app_module,
    app_callable,
    timeout_ms,
    scope,
    req_body_ch,
    resp_body_ch,
    body_ref,           %% Reference for async body reading
    has_body,
    headers_sent = false,
    handler_state       %% Original handler state from init
}).

-define(CHUNK_COALESCE_SIZE, 4096).
-define(CHUNK_COALESCE_TIMEOUT, 1).

%%% ============================================================================
%%% Cowboy Loop Handler Callbacks
%%% ============================================================================

init(Req, HandlerState) ->
    ReqInfo = build_request_info(Req),
    ReqInfo1 = hornbeam_http_hooks:run_on_request(ReqInfo),

    AppModule = maps:get(app_module, HandlerState),
    AppCallable = maps:get(app_callable, HandlerState),
    TimeoutMs = maps:get(timeout, HandlerState, 30000),

    %% Build ASGI scope
    Scope = hornbeam_request:build_asgi_scope(Req, HandlerState),

    %% Create channels
    {ok, ReqBodyCh} = py_byte_channel:new(),
    {ok, RespBodyCh} = py_byte_channel:new(),

    %% Check if request has a body
    Method = cowboy_req:method(Req),
    ContentLength = get_content_length(Req),
    HasBody = has_request_body(Method, ContentLength),

    %% Start async body reading if there's a body
    BodyRef = case HasBody of
        true ->
            Ref = make_ref(),
            %% Start async body reading - Cowboy will send us messages
            cowboy_req:cast({read_body, self(), Ref, auto, infinity}, Req),
            Ref;
        false ->
            %% No body - close channel immediately
            py_byte_channel:close(ReqBodyCh),
            undefined
    end,

    %% Create Python task
    _TaskRef = py_event_loop_pool:create_task(
        <<"hornbeam_asgi_loop">>, <<"handle_asgi_loop">>,
        [self(), AppModule, AppCallable, Scope, ReqBodyCh, RespBodyCh]),

    State = #state{
        req_info = ReqInfo1,
        app_module = AppModule,
        app_callable = AppCallable,
        timeout_ms = TimeoutMs,
        scope = Scope,
        req_body_ch = ReqBodyCh,
        resp_body_ch = RespBodyCh,
        body_ref = BodyRef,
        has_body = HasBody,
        handler_state = HandlerState
    },

    %% Return cowboy_loop to enable loop handler
    {cowboy_loop, Req, State, TimeoutMs}.

%% Handle async body chunks from Cowboy
info({request_body, Ref, nofin, Data}, Req, #state{body_ref = Ref, req_body_ch = Ch} = State) ->
    %% More body data coming - push to channel
    ok = push_to_channel(Ch, Data),
    %% Request more data
    cowboy_req:cast({read_body, self(), Ref, auto, infinity}, Req),
    {ok, Req, State};

info({request_body, Ref, fin, _BodyLen, Data}, Req, #state{body_ref = Ref, req_body_ch = Ch} = State) ->
    %% Final chunk - push and close channel
    case Data of
        <<>> -> ok;
        _ -> ok = push_to_channel(Ch, Data)
    end,
    py_byte_channel:close(Ch),
    {ok, Req, State#state{body_ref = undefined}};

%% Handle response headers from Python
info({<<"headers">>, StatusCode, Headers}, Req, State) ->
    SafeHeaders = filter_hop_by_hop(Headers),
    CowboyHeaders = convert_headers(SafeHeaders),
    Req2 = cowboy_req:stream_reply(StatusCode, CowboyHeaders, Req),
    %% Start draining response channel
    self() ! drain_response,
    {ok, Req2, State#state{headers_sent = true}};

%% Handle early hints from Python
info({<<"early_hints">>, Headers}, Req, State) ->
    HintHeaders = convert_headers(Headers),
    Req2 = cowboy_req:inform(103, HintHeaders, Req),
    {ok, Req2, State};

%% Drain response body channel
info(drain_response, Req, #state{resp_body_ch = RespBodyCh, handler_state = HandlerState} = State) ->
    case drain_response_chunk(RespBodyCh) of
        {ok, Data} ->
            ok = cowboy_req:stream_body(Data, nofin, Req),
            self() ! drain_response,
            {ok, Req, State};
        {error, closed} ->
            ok = cowboy_req:stream_body(<<>>, fin, Req),
            {stop, Req, HandlerState};
        {error, timeout} ->
            %% No data yet, schedule retry
            erlang:send_after(?CHUNK_COALESCE_TIMEOUT, self(), drain_response),
            {ok, Req, State}
    end;

%% Handle error from Python
info({<<"error">>, Reason}, Req, #state{req_info = ReqInfo, handler_state = HandlerState} = State) ->
    maybe_close_channel(State#state.req_body_ch),
    maybe_close_channel(State#state.resp_body_ch),
    {StatusCode, Body} = hornbeam_http_hooks:run_on_error(Reason, ReqInfo),
    Req2 = cowboy_req:reply(StatusCode,
                            #{<<"content-type">> => <<"text/plain">>},
                            Body, Req),
    {stop, Req2, HandlerState};

%% Handle async task completion
info({async_result, _Ref, {ok, _}}, Req, State) ->
    %% Task completed, continue draining response
    {ok, Req, State};

info({async_result, _Ref, {error, Reason}}, Req, #state{req_info = ReqInfo, handler_state = HandlerState} = State) ->
    maybe_close_channel(State#state.req_body_ch),
    maybe_close_channel(State#state.resp_body_ch),
    {StatusCode, Body} = hornbeam_http_hooks:run_on_error(Reason, ReqInfo),
    Req2 = cowboy_req:reply(StatusCode,
                            #{<<"content-type">> => <<"text/plain">>},
                            Body, Req),
    {stop, Req2, HandlerState};

%% Handle timeout
info(timeout, Req, #state{req_info = ReqInfo, handler_state = HandlerState} = State) ->
    maybe_close_channel(State#state.req_body_ch),
    maybe_close_channel(State#state.resp_body_ch),
    {StatusCode, Body} = hornbeam_http_hooks:run_on_error(timeout, ReqInfo),
    Req2 = cowboy_req:reply(StatusCode,
                            #{<<"content-type">> => <<"text/plain">>},
                            Body, Req),
    {stop, Req2, HandlerState};

%% Unknown message
info(_Msg, Req, State) ->
    {ok, Req, State}.

terminate(_Reason, _Req, _State) ->
    ok.

%%% ============================================================================
%%% Internal Functions
%%% ============================================================================

push_to_channel(Channel, Data) ->
    case py_byte_channel:send(Channel, Data) of
        ok -> ok;
        busy ->
            %% Channel full - wait a bit and retry
            timer:sleep(1),
            push_to_channel(Channel, Data);
        {error, closed} ->
            ok
    end.

drain_response_chunk(RespBodyCh) ->
    py_byte_channel:recv(RespBodyCh, ?CHUNK_COALESCE_TIMEOUT).

maybe_close_channel(undefined) -> ok;
maybe_close_channel(Channel) ->
    try
        case py_byte_channel:info(Channel) of
            #{closed := true} -> ok;
            _ ->
                catch py_byte_channel:close(Channel),
                ok
        end
    catch
        _:_ -> ok
    end.

%% @private
get_content_length(Req) ->
    case cowboy_req:header(<<"content-length">>, Req) of
        undefined -> undefined;
        CLBin ->
            try binary_to_integer(CLBin)
            catch _:_ -> undefined
            end
    end.

%% @private
has_request_body(<<"GET">>, undefined) -> false;
has_request_body(<<"HEAD">>, undefined) -> false;
has_request_body(<<"DELETE">>, undefined) -> false;
has_request_body(<<"OPTIONS">>, undefined) -> false;
has_request_body(_, 0) -> false;
has_request_body(_, _) -> true.

%% @private
build_request_info(Req) ->
    #{
        method => cowboy_req:method(Req),
        path => cowboy_req:path(Req),
        query_string => cowboy_req:qs(Req),
        headers => cowboy_req:headers(Req),
        host => cowboy_req:host(Req),
        port => cowboy_req:port(Req),
        scheme => cowboy_req:scheme(Req),
        peer => cowboy_req:peer(Req)
    }.

%% @private
filter_hop_by_hop(Headers) ->
    HopByHop = [<<"connection">>, <<"keep-alive">>, <<"proxy-authenticate">>,
                <<"proxy-authorization">>, <<"te">>, <<"trailers">>,
                <<"transfer-encoding">>, <<"upgrade">>],
    lists:filter(fun(Header) ->
        Name = case Header of
            [N, _] -> N;
            {N, _} -> N
        end,
        LowerName = string:lowercase(to_binary(Name)),
        not lists:member(LowerName, HopByHop)
    end, Headers).

%% @private
convert_headers(Headers) ->
    lists:foldl(fun(Header, Acc) ->
        case Header of
            [Name, Value] ->
                Acc#{to_lower_binary(Name) => to_binary(Value)};
            {Name, Value} ->
                Acc#{to_lower_binary(Name) => to_binary(Value)};
            _ ->
                Acc
        end
    end, #{}, Headers).

to_binary(V) when is_binary(V) -> V;
to_binary(V) when is_list(V) -> list_to_binary(V);
to_binary(V) when is_atom(V) -> atom_to_binary(V, utf8);
to_binary(V) -> iolist_to_binary(io_lib:format("~p", [V])).

to_lower_binary(V) when is_binary(V) -> string:lowercase(V);
to_lower_binary(V) when is_list(V) -> string:lowercase(list_to_binary(V));
to_lower_binary(V) when is_atom(V) -> string:lowercase(atom_to_binary(V, utf8));
to_lower_binary(V) -> string:lowercase(to_binary(V)).
